src/experimental/tools/videoThumbnailLoader/thumbnail_loader.ts

```diff
@@ -0,0 +1,261 @@
+/**
+ * Copyright 2015 CANAL+ Group
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import pinkie from "pinkie";
+import {
+  combineLatest as observableCombineLatest,
+  EMPTY,
+  Observable,
+  of as observableOf,
+} from "rxjs";
+import {
+  catchError,
+  mergeMap,
+  tap,
+} from "rxjs/operators";
+import Manifest from "../../../manifest";
+import {
+  disposeSourceBuffer,
+  getSourceBuffer$,
+} from "./get_source_buffer";
+import log from "./log";
+import {
+  loadInitThumbnail,
+  loadThumbnails
+} from "./segment_utils";
+import {
+  getThumbnailTrack,
+  getWantedThumbnails,
+  IThumbnail,
+  IThumbnailTrack,
+} from "./thumbnail_track_utils";
+
+const PPromise = typeof Promise === "function" ? Promise :
+                                                 pinkie;
+
+interface IJob { thumbnailTrack: IThumbnailTrack;
+                 thumbnails: IThumbnail[];
+                 stop: () => void; }
+
+// Returned error when rejecting
+class VideoThumbnailLoaderError extends Error {
+  public readonly name : "VideoThumbnailLoaderError";
+  public readonly message : string;
+  public readonly code : string;
+
+  /**
+   * @param {string} code
+   * @param {string} reason
+   * @param {Boolean} fatal
+   */
+  constructor(code : string, message : string) {
+    super();
+    Object.setPrototypeOf(this, VideoThumbnailLoaderError.prototype);
+    this.name = "VideoThumbnailLoaderError";
+    this.code = code;
+    this.message = message;
+  }
+}
+
+/**
+ * This tool, as a supplement to the RxPlayer, intent to help creating thumbnails
+ * from a video source.
+ *
+ * The tools will extract a "thumbnail track" either from a video track (whose light
+ * chunks are adapted from such use case) or direclty from the media content.
+ */
+export default class VideoThumbnailLoader {
+  private readonly _thumbnailVideoElement: HTMLVideoElement;
+
+  private _manifest: Manifest;
+  private _currentJob?: IJob;
+
+  constructor(videoElement: HTMLVideoElement,
+              manifest: Manifest) {
+    this._thumbnailVideoElement = videoElement;
+    this._manifest = manifest;
+  }
+
+  /**
+   * Set time of thumbnail video media element :
+   * - Remove buffer when too much buffered data
+   * - Search for thumbnail track element to display
+   * - Load data
+   * - Append data
+   * Resolves when time is set.
+   * @param {number} time
+   * @returns {Promise}
+   */
+  setTime(time: number): Promise<unknown> {
```

Comment:
Why Promise<unknown>?

Reply:
Sorry github formatting broke my previous comment which made no sense:

Why `Promise<unknown>`?

The `unknown` is the important part.